### ISSUE TITLE
Annotation based error-decoder

### ISSUE THREAD

**Issue by saintf** [OpenFeign/feign#489]
Hello - I looked around a while ago for an annotation based error decoder and couldn't really find one - so I wanted to see the apetite for one?

If it sounds interesting, I can clone the repo, create a new module under it (thoughts on naming?) and put something good in there for review?

Basically - it would be managed via 2 annotations:
`@ErrorDecoding` which would take a few parameters, including a second annotation named something to the effect of `@ErrorCodes` (the idea being that multiple error codes could still map to the same annotation).

So basically, in your interface you'd annotate your method (or class) with something like:
`@ErrorDecoding { @ErrorCodes { [401, 403], mapTo = UnauthenticatedOrUnauthorizedException.class }}`

Then - you'd just pass it into the Feign.Builder via the ErrorDecoder as a new AnnotationBasedErrorDecoder or something to that effect. 

Thoughts?

**Comment by arielsegura** [comment]
+1

**Comment by spencergibb** [comment]
@arielsegura please use the "Add your reaction" widget rather than a comment.

**Comment by adriancole** [comment]
you could do something like this, since you have access to `String methodKey`. For example, you can make your own contract that builds a map of methodKey to error decoder. the value would be the thing that's parsed from annotations. That map is as a dispatch error decoder, when no key is present, it means no annotations were parsed, so whatever the "normal" error decoder is invoked.

**Comment by saintf** [comment]
ok - so I've just committed my first iteration of this into a branch and created a PR.

https://github.com/OpenFeign/feign/pull/509

Clearly I don't know github well enough :) I created it so we could review the code but now see the commit on the branch. 

Can I get a first set of eyes for comment on the direction I'm taking and I can take it from there and maybe get something nice in? :)

**Comment by saintf** [comment]
right - last comment - I couldn't create a branch in this repo (and hence couldn't commit to it either), and I went through some hoops committing onto my forked one (which deleted it initially/etc - most of it likely dude to user error).

I've recreated my fork, committed the changes there. Let me know if you can't see it. I can keep working from it and then raise a PR when we're good for me to merge into here. 

Thanks,
-Fernando

**Comment by igorbljahhin** [comment]
+1

**Comment by saintf** [comment]
Cleaning up left behind ticket. The decoder was developed/added way back when.

Can be found here:
https://github.com/OpenFeign/feign-annotation-error-decoder
